### PR TITLE
libgit2: add version 1.7.2, support conan v2

### PR DESCRIPTION
**Pull request by toge**
Specify library name and version:  **libgit2/***

---

- [x] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [x] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by whaliendev** [comment]
Hey, @toge I wonder why boost is needed to built this library? 
when I run `fd -i "boost"`, it shows nothing. It's very strange that libgit2 depends on boost.

**Comment by toge** [comment]
Um... openssl/3.1.2?
```
WARN: libssh2/1.11.0: requirement openssl/[>=1.1 <4] overridden by libgit2/1.7.0 to openssl/3.1.2 
ERROR: Failed requirement 'openssl/3.1.2' from 'libssh2/1.11.0'
ERROR: Unable to find 'openssl/3.1.2' in remotes
```

File: recipes/libgit2/0.27.x/conanfile.py [hunk 1]
```diff
@@ -38,43 +41,42 @@ class LibGit2Conan(ConanFile):
...
     def requirements(self):
-        self.requires("zlib/1.2.12")
+        self.requires("zlib/1.2.13")
         self.requires("http_parser/2.9.4")
         if self.options.with_libssh2:
-            self.requires("libssh2/1.10.0")
+            self.requires("libssh2/1.11.0")
         if self.settings.os != "Windows":
-            self.requires("libcurl/7.83.1")
+            self.requires("libcurl/8.1.2")
```
Comment (uilianries) [inline recipes/libgit2/0.27.x/conanfile.py:73]:
```suggestion
            self.requires("libcurl/[>=X.YY <9]")
```

https://github.com/conan-io/conan-center-index/blob/master/docs/adding_packages/dependencies.md#version-ranges

File: recipes/libgit2/0.27.x/conanfile.py [hunk 2]
```diff
@@ -38,43 +41,42 @@ class LibGit2Conan(ConanFile):
...
+    def layout(self):
+        cmake_layout(self, src_folder="src")
 
     @property
     def _need_openssl(self):
         return "openssl" in (self.options.with_https, self.options.with_sha1)
 
     def requirements(self):
-        self.requires("zlib/1.2.12")
+        self.requires("zlib/1.2.13")
```
Comment (uilianries) [inline recipes/libgit2/0.27.x/conanfile.py:68]:
```suggestion
        self.requires("zlib/[>=1.X.Y <2]")
```

https://github.com/conan-io/conan-center-index/blob/master/docs/adding_packages/dependencies.md#version-ranges

File: recipes/libgit2/0.28.x/conanfile.py [hunk 3]
```diff
@@ -67,20 +67,20 @@ def _need_mbedtls(self):
         return "mbedtls" in (self.options.with_https, self.options.with_sha1)
 
     def requirements(self):
-        self.requires("zlib/1.2.12")
+        self.requires("zlib/1.2.13")
```
Comment (uilianries) [inline recipes/libgit2/0.28.x/conanfile.py:70]:
```suggestion
        self.requires("zlib/[>=1.X.Y <2]")
```

https://github.com/conan-io/conan-center-index/blob/master/docs/adding_packages/dependencies.md#version-ranges

**Review by uilianries — COMMENTED** [review]
Looking good, only small suggestion related to range versions.

File: recipes/libgit2/0.28.x/conanfile.py (hunk 3 above)
Comment (toge) [reply recipes/libgit2/0.28.x/conanfile.py:70]: @uilianries 
Thanks!
I fixed.

**Review by uilianries — APPROVED** [review]
Arigatou!!

File: recipes/libgit2/0.27.x/conanfile.py [hunk 4]
```diff
@@ -38,43 +41,42 @@ class LibGit2Conan(ConanFile):
...
+    def layout(self):
+        cmake_layout(self, src_folder="src")
 
     @property
     def _need_openssl(self):
         return "openssl" in (self.options.with_https, self.options.with_sha1)
 
     def requirements(self):
-        self.requires("zlib/1.2.12")
+        self.requires("zlib/[>=1.X.Y <2]")
```
Comment (uilianries) [inline recipes/libgit2/0.27.x/conanfile.py:68]:
```suggestion
        self.requires("zlib/[>=1.2.12 <2]")
```

File: recipes/libgit2/0.27.x/conanfile.py [hunk 5]
```diff
@@ -38,43 +41,42 @@ class LibGit2Conan(ConanFile):
...
         self.requires("http_parser/2.9.4")
         if self.options.with_libssh2:
-            self.requires("libssh2/1.10.0")
+            self.requires("libssh2/1.11.0")
         if self.settings.os != "Windows":
-            self.requires("libcurl/7.83.1")
+            self.requires("libcurl/[>=X.YY <9]")
```
Comment (uilianries) [inline recipes/libgit2/0.27.x/conanfile.py:73]:
```suggestion
            self.requires("libcurl/[>=7.83 <9]")
```

**Comment by uilianries** [comment]
@toge I created the PR https://github.com/conan-io/conan-center-index/pull/19672 to fix the current conflict.

**Comment by RubenRBS** [comment]
The bot did not want to post the errors, but v2 is failing because of a conflict in zlib, will get fixed once we continue with the port, sorry about the delay

**Review by jwillikers — APPROVED** [review]

**Comment by AlexandrePTJ** [comment]
Hi, is there any status on this PR ?
We do rely on this package, and this is the last one before we can completly migrate to conan 2.

**Comment by uilianries** [comment]
I just started new internal build to generate de missing `libssh2/1.11.0` package with `openssl/3.2.0`

**Comment by toge** [comment]
@uilianries 
There seem to be still missing dependencies.
Cloud you investigate them?

**Comment by uilianries** [comment]
I just started an internal build to generate 'libssh2/1.11.0', 'openssl/3.2.1'.

**Comment by uilianries** [comment]
@toge Sorry the incident, I guess it failed during the tapaholes. I just pushed to be built it again. Thank you for reporting!

File: recipes/libgit2/0.27.x/conanfile.py [hunk 6]
```diff
@@ -1,10 +1,13 @@
...
+from conan.tools.cmake import CMake, CMakeDeps, CMakeToolchain, cmake_layout
+from conan.tools.apple import is_apple_os
 import os
 
-required_conan_version = ">=1.45.0"
+
+required_conan_version = ">=1.53.0"
 
 
 class LibGit2Conan(ConanFile):
```
Comment (valgur) [inline recipes/libgit2/0.27.x/conanfile.py:13]: `package_type` is missing.

File: recipes/libgit2/0.27.x/conandata.yml
Comment (valgur) [inline recipes/libgit2/0.27.x/conandata.yml:1]: Maybe the `0.27` recipe can be dropped altogether.

File: recipes/libgit2/0.27.x/conanfile.py [hunk 7]
```diff
@@ -103,49 +104,32 @@ def source(self):
...
-
+    def generate(self):
+        tc = CMakeToolchain(self)
+        tc.variables["THREADSAFE"] = self.options.threadsafe
+        tc.variables["USE_SSH"] = self.options.with_libssh2
+        tc.variables["USE_ICONV"] = self.options.get_safe("with_iconv", False)
+        tc.variables["USE_HTTPS"] = self._cmake_https[str(self.options.with_https)]
+        tc.variables["SHA1_BACKEND"] = self._cmake_sha1[str(self.options.with_sha1)]
+        tc.variables["BUILD_CLAR"] = False
+        tc.variables["BUILD_EXAMPLES"] = False
```
Comment (valgur) [inline recipes/libgit2/0.27.x/conanfile.py:115]: Need to set `CMP0077` to `NEW` for `tc.variables`.

File: recipes/libgit2/all/conanfile.py [hunk 8]
```diff
@@ -135,47 +134,44 @@ def source(self):
...
+        tc.variables["USE_ICONV"] = self.options.get_safe("with_iconv", False)
+        tc.variables["USE_HTTPS"] = self._cmake_https[str(self.options.with_https)]
+        tc.variables["USE_SHA1"] = self._cmake_sha1[str(self.options.with_sha1)]
+        if Version(self.version) >= "1.4.0":
+            tc.variables["BUILD_TESTS"] = False
+        tc.variables["BUILD_CLAR"] = False
+        tc.variables["BUILD_CLI"] = False
+        tc.variables["BUILD_EXAMPLES"] = False
+        tc.variables["USE_HTTP_PARSER"] = "system"
+        tc.variables["REGEX_BACKEND"] = self.options.with_regex
```
Comment (valgur) [inline recipes/libgit2/all/conanfile.py:150]: Needs a `CMP0077` as well.

File: recipes/libgit2/0.28.x/conanfile.py [hunk 9]
```diff
@@ -111,45 +110,32 @@ def source(self):
...
-
-        cmake.definitions["BUILD_CLAR"] = False
-        cmake.definitions["BUILD_EXAMPLES"] = False
-
+    def generate(self):
+        tc = CMakeToolchain(self)
+        tc.variables["THREADSAFE"] = self.options.threadsafe
+        tc.variables["USE_SSH"] = self.options.with_libssh2
+        tc.variables["USE_ICONV"] = self.options.get_safe("with_iconv", False)
+        tc.variables["USE_HTTPS"] = self._cmake_https[str(self.options.with_https)]
```
Comment (valgur) [inline recipes/libgit2/0.28.x/conanfile.py:118]: The `False` value in the dict is not a string.

File: recipes/libgit2/all/conanfile.py [hunk 10]
```diff
@@ -135,47 +134,44 @@ def source(self):
...
-        cmake.definitions["USE_HTTP_PARSER"] = "system"
-
-        cmake.definitions["REGEX_BACKEND"] = self.options.with_regex
-
+    def generate(self):
+        tc = CMakeToolchain(self)
+        tc.variables["THREADSAFE"] = self.options.threadsafe
+        tc.variables["USE_SSH"] = self.options.with_libssh2
+        tc.variables["USE_ICONV"] = self.options.get_safe("with_iconv", False)
+        tc.variables["USE_HTTPS"] = self._cmake_https[str(self.options.with_https)]
```
Comment (valgur) [inline recipes/libgit2/all/conanfile.py:142]: Same, `False` should be a string in the dict.

**Review by valgur — CHANGES_REQUESTED** [review]
Some minor tweaks are required. I would drop `0.27` as well. Maybe `0.28` also, but it's the latest v0.x version on CCI so it's still marginally useful in that regard, possibly.

**Comment by RubenRBS** [comment]
Hi @toge - I'm now trying to come back to this PR, sorry about the delay.

My one main question is if you have any insight into keeping the old 0.x releases, I think we could simplify a lot if only the `all` recipe would be kept! :)

**Comment by toge** [comment]
@RubenRBS 
I agree with you.
I will try to remove 0.x releases.

**Comment by RubenRBS** [comment]
Hi! I went ahead and removed them and updated the recipe to use version ranges, sorry I didnt notify you sooner! :)

File: recipes/libgit2/all/conanfile.py [hunk 11]
```diff
@@ -133,49 +132,49 @@ def source(self):
...
+        tc.variables["BUILD_EXAMPLES"] = False
+        tc.variables["USE_HTTP_PARSER"] = "system"
+        tc.variables["REGEX_BACKEND"] = self.options.with_regex
         if is_msvc(self):
-            cmake.definitions["STATIC_CRT"] = is_msvc_static_runtime(self)
-
-        cmake.configure()
-        return cmake
+            tc.variables["STATIC_CRT"] = is_msvc_static_runtime(self)
+        # REGEX_BACKEND is SET(), avoid options overriding it
```
Comment (uilianries) [inline recipes/libgit2/all/conanfile.py:154]: Thank you for commenting the motivation! It helps a lot!

File: recipes/libgit2/all/patches/1.7.2-0001-fix-cmake.patch [hunk 12]
```diff
@@ -0,0 +1,70 @@
+diff --git a/cmake/SelectHTTPParser.cmake b/cmake/SelectHTTPParser.cmake
+index 955aea3..ff94eb8 100644
+--- a/cmake/SelectHTTPParser.cmake
++++ b/cmake/SelectHTTPParser.cmake
+@@ -1,10 +1,10 @@
+ # Optional external dependency: http-parser
+ if(USE_HTTP_PARSER STREQUAL "system")
+-	find_package(HTTPParser)
```
Comment (uilianries) [inline recipes/libgit2/all/patches/1.7.2-0001-fix-cmake.patch:8]: We could use CMakeDeps with set_property to change the cmake file name, but variables are still not possible to change, so this change looks okay.

**Review by RubenRBS — APPROVED** [review]
Thanks a lot @toge, sorry this one took so long to get merged :)

**Review by uilianries — APPROVED** [review]

**Pull request merged**